modules/apps/portal-odata/portal-odata-api/src/main/java/com/liferay/portal/odata/filter/expression/factory/ExpressionFactory.java

```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portal.odata.filter.expression.factory;
+
+import com.liferay.portal.odata.filter.expression.BinaryExpression;
+import com.liferay.portal.odata.filter.expression.CollectionPropertyExpression;
+import com.liferay.portal.odata.filter.expression.ComplexPropertyExpression;
+import com.liferay.portal.odata.filter.expression.Expression;
+import com.liferay.portal.odata.filter.expression.LambdaFunctionExpression;
+import com.liferay.portal.odata.filter.expression.LambdaVariableExpression;
+import com.liferay.portal.odata.filter.expression.ListExpression;
+import com.liferay.portal.odata.filter.expression.LiteralExpression;
+import com.liferay.portal.odata.filter.expression.MemberExpression;
+import com.liferay.portal.odata.filter.expression.MethodExpression;
+import com.liferay.portal.odata.filter.expression.NavigationPropertyExpression;
+import com.liferay.portal.odata.filter.expression.PrimitivePropertyExpression;
+import com.liferay.portal.odata.filter.expression.PropertyExpression;
+import com.liferay.portal.odata.filter.expression.UnaryExpression;
+
+import java.util.List;
+
+/**
+ * @author Carlos Correa
+ */
+public interface ExpressionFactory {
```

Comment:
This is totally an open question. 
Do we want to allow the creation of Expressions from outside of the OData module?

My fear (and I could be wrong) is that if we allow this, other modules will have a dependency on OData modules, which could end in a dependency on the REST layer.
Something similar to what we had with Vulcan when other modules were using our TranformUtil.

As I said, I could be totally wrong. It is just a thought that comes to my mind reading the description of this commit.

Reply:
We have the following options:

- Merge the request filter with the API endpoint filter as Strings
- Merge the request filter with the API endpoint filter as Expression
- Merge the request filter with the API endpoint filter as Predicate

As we have to validate the request filter, we need to transform it at least to an Expression, because we delegate the validation given an EntityModel. As we need to translate the API Builder property names into the internal names, we should implement for the first option (String option) a Visitor that transforms the already validated Expression into String (because there is no way OOTB in the olingo implementation).

The second option is the one that we implemented.

With the third one, we would be including more Predicates in the REST Layer of the API Builder. That is another option, would you prefer that option or Expressions?

I would like to know your opinion @4lejandrito  too. For me, both options are OK, but I thought we wanted to get rid of the Predicates in the REST layer :thinking:

Reply:
1. Would impact performance because the filter needs to be parsed back and forth on each request.
2. I still prefer this one.
3. This would leak the Predicate (DB) to the outer world, which is really an implementation detail. The interface should be odata I think.